metals/src/main/scala/scala/meta/internal/metals/ScalafixProvider.scala

```diff
@@ -0,0 +1,154 @@
+    val unsavedFile = file.toInputFromBuffers(buffers)
+    if (md5(fileInput.text) != md5(unsavedFile.text)) {
+      scribe.info(s"Organize imports requires saving the file first")
+      languageClient.showMessage(
+        MessageType.Warning,
+        s"Save ${file.toNIO.getFileName} before organizing imports"
+      )
+      Nil
+    } else {
+      val baseMatcher =
+        FileSystems.getDefault.getPathMatcher("glob:**.{scala,sbt}")
+      if (baseMatcher.matches(file.toNIO)) {
+        val scalafixConfPath = workspace.resolve(scalafixConfigPath)
+        val scalafixConf =
+          if (scalafixConfPath.isFile) Some(scalafixConfPath.toNIO)
+          else None
+        val scalafixEvaluation = for {
+          (scalaVersion, scalaBinaryVersion, classPath) <-
+            scalaVersionAndClasspath(file)
+          api <- getOrUpdateScalafixCache(scalaBinaryVersion)
+          scalafixArgs = configureApi(api, scalaVersion, classPath)
+          urlClassLoaderWithExternalRule = embedded.organizeImportsClassloader(
+            scalaBinaryVersion,
+            api.getClass.getClassLoader
+          )
+        } yield {
+          val scalacOption =
+            if (scalaBinaryVersion == "2.13") "-Wunused:imports"
+            else "-Ywarn-unused-import"
+
+          scalafixArgs
+            .withToolClasspath(urlClassLoaderWithExternalRule)
+            .withConfig(scalafixConf.asJava)
+            .withRules(List(organizeImportRuleName).asJava)
+            .withPaths(List(file.toNIO).asJava)
+            .withSourceroot(workspace.toNIO)
+            .withScalacOptions(Collections.singletonList(scalacOption))
+            .evaluate()
+        }
+
+        val newFileContentOpt = scalafixEvaluation
+          .flatMap(result => result.getFileEvaluations.headOption)
+          .flatMap(_.previewPatches().asScala)
+        newFileContentOpt.map(textEditsFrom(_, fileInput)).getOrElse(Nil)
+      } else {
+        scribe.info(
+          s"""|Could not organize import for ${file.toNIO.getFileName}. Should end with .scala or .sbt"""
+        )
+        Nil
+      }
+    }
+  }
+
+  private def textEditsFrom(
+      newFileContent: String,
+      input: Input
+  ): List[l.TextEdit] = {
+    val fullDocumentRange = Position.Range(input, 0, input.chars.length).toLSP
+    if (newFileContent != input.text) {
+      List(new l.TextEdit(fullDocumentRange, newFileContent))
+    } else {
+      Nil
+    }
+  }
+
+  private def scalaVersionAndClasspath(
+      file: AbsolutePath
+  ): Option[(ScalaVersion, ScalaBinaryVersion, util.List[Path])] =
+    for {
+      identifier <- buildTargets.inverseSources(file)
+      scalacOptions <- buildTargets.scalacOptions(identifier)
+      scalaBuildTarget <- buildTargets.scalaInfo(identifier)
+      scalaVersion = scalaBuildTarget.getScalaVersion
+      semanticdbTarget = scalacOptions.targetroot(scalaVersion).toNIO
+      scalaBinaryVersion = scalaBuildTarget.getScalaBinaryVersion
```

Comment:
We should print a warning in case of Scala 3 versions, otherwise I think it will throw errors out.

Reply:
Forgot about scala 3..

Reply:
It wont throw an exception, since it's cached in a try, but I will add an error message